### PR TITLE
Added support for instance Id logic in k8s environment

### PR DESCRIPTION
**Pull request by chtyim**
- Works for all supported resources type (Deployment, StatefulSet, Job)
- Handling of scaling up and down
- Handling of pod restart correctly
  - Old logic can create infinite loop

### CODE REVIEW

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillContext.java [hunk 1]
```diff
@@ -211,101 +228,304 @@ public String getUID() {
...
+        }
+        count = value.getAsInt();
       }
 
-      @Override
-      public void resourceAdded(Object resource) {
-        resourceModified(resource);
+      int oldInstances = instances.getAndSet(count);
+      if (oldInstances != count) {
+        LOG.debug("Number of instances change to {}", count);
```
Comment (CuriousVini) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillContext.java]: should we add `Number of instances changed from {} to {}`?

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillContext.java [hunk 2]
```diff
@@ -211,101 +228,304 @@ public String getUID() {
...
+      }
+
+      // If it is not StatefulSet or failed to extract instanceId from the pod name, use the assignment logic
+      performAssignInstanceId(api);
+    }
+
+    /**
+     * Performs the instance Id assignment logic using annotations on the owner object.
+     * The logic involves setting an annotation with key {@link #INSTANCES_ASSIGNMENT}, and value is a Json map
+     * that stores the associate of instance Id to pod name ({@code instanceId -> Pod Name}).
```
Comment (CuriousVini) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillContext.java]: nit: `stores association of ..`

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillContext.java [hunk 3]
```diff
@@ -211,101 +228,304 @@ public String getUID() {
...
+     * Each of the pod instances will race to update the assignment map, which we rely on the compare and set property
+     * of the k8s replace API to guarantee a consistent view across all pods.
```
Comment (CuriousVini) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillContext.java:399]: should we say `put API` instead?

**Review by CuriousVini — COMMENTED** [review]
Thanks! LGTM 👍🏼

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillContext.java (hunk 1 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillContext.java]: Sure

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillContext.java (hunk 2 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillContext.java]: Fixed

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillContext.java (hunk 3 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillContext.java:399]: The http method is PUT, but the k8s api call it Replace. See https://v1-18.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#-strong-write-operations-replicaset-v1-apps-strong-

**Review by CuriousVini — APPROVED** [review]
Thanks, LGTM 👍🏼

**Pull request merged**